src/execution_plan/optimizations/optimize_cartesian_product.c

```diff
@@ -0,0 +1,140 @@
+	OpFilter **filter_ops = _locate_filters(cp);
+	uint filter_count = array_len(filter_ops);
+	/* This array will hold filters which are possible to re position in the execution plan,
+	 * but during the optimization construction they will not cause this optimization to execute. */
+	OpFilter **pending_filters = array_new(OpFilter *, filter_count);
+	OpBase *pending_filters_lookup_root = cp;
+
+	// For each stream joined by the Cartesian product, collect all entities the stream resolves.
+	int stream_count = cp->childCount;
+	rax *stream_entities[stream_count];
+	build_streams_from_op(cp, stream_entities, stream_count);
+
+	for(uint i = 0; i < filter_count; i ++) {
+		// Try to create a new dual-branched cartesian product, followed by the current filter.
+		OpFilter *filter_op = filter_ops[i];
+
+		/* A new cartesian product can be generated if both sides of the filter can be fully and
+		 * separately resolved by exactly two child streams. */
+		FT_FilterNode *f = filter_op->filterTree;
+
+		/* Make sure LHS of the filter is resolved by a stream. */
+		AR_ExpNode *lhs = f->pred.lhs;
+		uint lhs_resolving_stream = relate_exp_to_stream(lhs, stream_entities, stream_count);
+		if(lhs_resolving_stream == NOT_RESOLVED) continue;
+		/* Make sure RHS of the filter is resolved by a stream. */
+		AR_ExpNode *rhs = f->pred.rhs;
+		uint rhs_resolving_stream = relate_exp_to_stream(rhs, stream_entities, stream_count);
+		if(rhs_resolving_stream == NOT_RESOLVED) continue;
+
+		// This stream is solved by a single cartesian product child and needs to be propogate later.
+		if(lhs_resolving_stream == rhs_resolving_stream) {
+			pending_filters = array_append(pending_filters, filter_op);
+			continue;
+		}
+
+		// Retrieve the relevent branch roots.
+		OpBase *right_branch = cp->children[rhs_resolving_stream];
+		OpBase *left_branch = cp->children[lhs_resolving_stream];
+		// Detach the streams and completely remove the filter from the execution plan.
+		ExecutionPlan_DetachOp(right_branch);
+		ExecutionPlan_DetachOp(left_branch);
+		ExecutionPlan_RemoveOp(plan, (OpBase *)filter_op);
+		// Combine the streams as cartesian product and filter them.
+		OpBase *filtered_cp_branch = _chain_cp_and_filter(cp->plan, left_branch, right_branch,
+														  (OpBase *)filter_op);
+
+		if(cp->childCount == 0) {
+			// The entire Cartesian Product can be replaced with the join op.
+			ExecutionPlan_ReplaceOp(plan, cp, filtered_cp_branch);
+			OpBase_Free(cp);
+			pending_filters_lookup_root = filtered_cp_branch;
+			break;
+		} else {
+			// The Cartesian Product still has a child operation; introduce the new op as another child.
+			ExecutionPlan_AddOp(cp, filtered_cp_branch);
+			// Streams are no longer valid since cartesian product changed.
+			for(int j = 0; j < stream_count; j ++) raxFree(stream_entities[j]);
+			// Re-collect cartesian product streams.
+			stream_count = cp->childCount;
+			build_streams_from_op(cp, stream_entities, stream_count);
```

Comment:
We only need to rebuild streams if we still have filters to evaluate. Perhaps `if (i+1 == filter_count) break`?

Reply:
you merge streams (children) so the current number of streams and their structrure in not alligned with the current state of the execution plan.

Reply:
Only if the loop is going to execute again, though - if we've just evaluated the last filter, this is unnecessary effort.